### PR TITLE
callback after DigitalOutputDevice.blink ends

### LINKED ISSUE #827: callback after DigitalOutputDevice.blink ends
Currently there is no indication, when a DigitalOutputDevice stops blinking because n blinks are done or another state was set.
In order to build complex blinking sequences or simply to set the value after blinking a callback would be helpful.
This callback should be called with a parameter indicating whether the blinking stops after the cycle was completed or the cycle was terminated by another command.

### PR DESCRIPTION
**Pull request by MNeum**
close #827

enhance blink method in DigitalOutputDevice, PWMOutputDevice and RGBLED:
- additional parameter on_end: callback function to call after blink ended
- additional parameter end_state / end_value / end_color: will be set, if a blink sequence was finished uninterrupted

new blink_sequence method in DigitalOutputDevice, PWMOutputDevice and RGBLED runs blinking sequences

additional feature: wait delay is time corrected, so that multiple independently blinking devices stay in sync

### CODE REVIEW

**Comment by MNeum** [comment]
Looking through the travis build log, I see that the pypy3 build failed because travis was unable to download pypy3 archive

I hope, this is not serious...

File: gpiozero/output_devices.py [hunk 1]
```diff
@@ -236,7 +240,8 @@ def off(self):
         self._stop_blink()
         self._write(False)
 
-    def blink(self, on_time=1, off_time=1, n=None, background=True):
+    def blink(self, on_time=1, off_time=1, end_state=None,
```
Comment (lurch) [inline gpiozero/output_devices.py]: I think for backwards-compatibility reasons, we prefer to not reorder the args-list. So here you'd need to move `end_state` to after `background` .

File: gpiozero/output_devices.py [hunk 2]
```diff
@@ -324,7 +384,11 @@ class LED(DigitalOutputDevice):
         See :doc:`api_pins` for more information (this is an advanced feature
         which most users can ignore).
     """
-    pass
+    def __init__(self, pin=None, active_high=True, initial_value=False, pin_factory=None):
+        self._blink_thread = None
+        self._controller = None
+        super(LED, self).__init__(pin, active_high, initial_value, pin_factory=pin_factory)
```
Comment (lurch) [inline gpiozero/output_devices.py]: Why is this change needed, when `DigitalOutputDevice`'s `__init__` already sets `_blink_thread` and `_controller` to None? :man_shrugging:

File: gpiozero/output_devices.py [hunk 3]
```diff
@@ -517,6 +583,22 @@ def blink(
...
+        :param float off_value:
+            value during off_time.
+            Defaults to 0 (full off).
+
+        :type end_value: float or None
+        :param end_value:
+            value to set after a full uninterrupted blink.
+            Defaults to None, on which no value will be set.
+
+        :param int fps:
```
Comment (lurch) [inline gpiozero/output_devices.py]: Why add this as an extra parameter? Wouldn't it make more sense to just (internally) use `self.frequency` ?

**Comment by bennuttall** [comment]
Thanks for putting the effort in but I'm not convinced it's necessary. You could subclass to add this feature - I'd consider it API bloat.

File: gpiozero/output_devices.py [hunk 4]
```diff
@@ -526,41 +608,125 @@ def blink(
...
+        self._blink_thread.start()
+        if not background:
+            self._blink_thread.join()
+            self._blink_thread = None
+
+    @staticmethod
+    def interpolate_value(start_value, end_value, delta_t, t):
+        return (start_value * (delta_t - t) + end_value * t) / delta_t
+
+    def _calculate_sequence(self, on_value, fade_in_time, on_time, off_value, fade_out_time, off_time, fps):
```
Comment (lurch) [inline gpiozero/output_devices.py]: Could this also be a `@staticmethod` ?

File: gpiozero/output_devices.py [hunk 5]
```diff
@@ -889,23 +1055,44 @@ class RGBLED(SourceMixin, Device):
         See :doc:`api_pins` for more information (this is an advanced feature
         which most users can ignore).
 
-    .. _colorzero: https://colorzero.readthedocs.io/
+    .. _colorzero: https://colorzero.readtheocs.io/
```
Comment (lurch) [inline gpiozero/output_devices.py]: ???

**Comment by lurch** [comment]
Why is the `end_state` parameter necessary - surely it's easy enough to just set that from the callback function?
(I started reviewing this before realising how big the changes were, so I only got part-way through the review... :wink: )

File: gpiozero/output_devices.py (hunk 1 above)
Comment (MNeum) [reply gpiozero/output_devices.py]: I didn't think about that. No problem!

File: gpiozero/output_devices.py (hunk 3 above)
Comment (MNeum) [reply gpiozero/output_devices.py]: It would be possible to use self.frequency. I borrowed this from the _blink_device, where it was a default parameter, but never set on the call.
fps defines, how smooth the transition of the fade will be. 
A value of 25 means, that every 40 ms a new value will be set.  
On a PWM frequency of 100 Hz that means every 4th zero crossing the pulse width will be changed. That is a heavy load already. In most cases a lower value is enough for a smooth eye pleasing transition.

File: gpiozero/output_devices.py (hunk 4 above)
Comment (MNeum) [reply gpiozero/output_devices.py]: Yes, you are right.

File: gpiozero/output_devices.py (hunk 5 above)
Comment (MNeum) [reply gpiozero/output_devices.py]: Definitely a bad error...

**Comment by MNeum** [comment]
> Why is the `end_state` parameter necessary - surely it's easy enough to just set that from the callback function?
In my application this is a frequent used feature, whereas the callback is seldom needed (only for sync purposes). 
But you are right, end_state can be set by a callback and the end_set parameter could be removed.

**Comment by lurch** [comment]
> Maybe blinking functionality could be a Mixin?

I had that idea too a few years ago :wink:  I started working on it but then I ran into some problem (I can't remember now what it was - perhaps something to do with RGBLED) and I didn't have enough enthusiasm to pursue the idea further.

> the same problem I try to handle here ... complex rules for switching on leds ... sync several led timings

Sounds to me like you'd probably be better off just creating your own timing threads, and only using gpiozero to turn the LEDs `.on()` and `.off()` as needed, rather than trying to modify the (deliberately limited) `blink_thread` to do everything you need it to? :man_shrugging:

**Comment by MNeum** [comment]
Since there seems no desire to implement this in the core library, I suppose we can close this PR
